src/rpc/fam_rpc_service_impl.cpp

```diff
@@ -718,33 +739,99 @@ int Fam_Rpc_Service_Impl::deregister_memory(uint64_t regionId,
     uint64_t rwKey = generate_access_key(regionId, dataitemId, 1);
+    Fam_Region_Map_t *fiRegionMap;
 
-    pthread_mutex_lock(famOps->get_mr_lock());
-    auto rMr = fiMrs->find(rKey);
-    auto rwMr = fiMrs->find(rwKey);
-    if (rMr != fiMrs->end()) {
+    if (fiMrs == NULL)
+        fiMrs = famOps->get_fiMrs();
+
+    // Take read lock on fiMrs
+    pthread_rwlock_rdlock(famOps->get_mr_lock());
+    auto regionMrObj = fiMrs->find(regionId);
+    if (regionMrObj == fiMrs->end()) {
+        pthread_rwlock_unlock(famOps->get_mr_lock());
+        return 0;
+    } else {
+        fiRegionMap = regionMrObj->second;
+    }
+
+    // Take a writelock on fiRegionMap
+    pthread_rwlock_wrlock(&fiRegionMap->fiRegionLock);
+    // Release lock on fiMrs
+    pthread_rwlock_unlock(famOps->get_mr_lock());
+
+    auto rMr = fiRegionMap->fiRegionMrs->find(rKey);
+    auto rwMr = fiRegionMap->fiRegionMrs->find(rwKey);
+    if (rMr != fiRegionMap->fiRegionMrs->end()) {
         ret = fabric_deregister_mr(rMr->second);
         if (ret < 0) {
-            pthread_mutex_unlock(famOps->get_mr_lock());
+            pthread_rwlock_unlock(&fiRegionMap->fiRegionLock);
+            cout << "error: memory deregister failed" << endl;
             return ITEM_DEREGISTRATION_FAILED;
         }
-        fiMrs->erase(rMr);
+        fiRegionMap->fiRegionMrs->erase(rMr);
     }
 
-    if (rwMr != fiMrs->end()) {
+    if (rwMr != fiRegionMap->fiRegionMrs->end()) {
         ret = fabric_deregister_mr(rwMr->second);
         if (ret < 0) {
-            pthread_mutex_unlock(famOps->get_mr_lock());
+            pthread_rwlock_unlock(&fiRegionMap->fiRegionLock);
+            cout << "error: memory deregister failed" << endl;
             return ITEM_DEREGISTRATION_FAILED;
         }
-        fiMrs->erase(rwMr);
+        fiRegionMap->fiRegionMrs->erase(rwMr);
     }
 
-    pthread_mutex_unlock(famOps->get_mr_lock());
+    pthread_rwlock_unlock(&fiRegionMap->fiRegionLock);
     RPC_SERVICE_PROFILE_END_OPS(deregister_memory);
     return 0;
 }
 
+int Fam_Rpc_Service_Impl::deregister_region_memory(uint64_t regionId) {
+    int ret = 0;
+    Fam_Region_Map_t *fiRegionMap;
+
+    if (fiMrs == NULL)
+        fiMrs = famOps->get_fiMrs();
+
+    // Take write lock on fiMrs
+    pthread_rwlock_wrlock(famOps->get_mr_lock());
+
+    auto regionMrObj = fiMrs->find(regionId);
+    if (regionMrObj == fiMrs->end()) {
+        pthread_rwlock_unlock(famOps->get_mr_lock());
+        return 0;
+    } else {
+        fiRegionMap = regionMrObj->second;
+    }
+
+    // Take a writelock on fiRegionMap
+    pthread_rwlock_wrlock(&fiRegionMap->fiRegionLock);
+    // Remove region map from fiMrs
+    fiMrs->erase(regionMrObj);
+    // Release lock on fiMrs
+    pthread_rwlock_unlock(famOps->get_mr_lock());
+
+    // Unregister all dataItem memory from region map
+    for (auto mr : *(fiRegionMap->fiRegionMrs)) {
+        ret = fabric_deregister_mr(mr.second);
+        if (ret < 0) {
```

Comment:
we should not stop here. We should go ahead and remove rest of the entries. And again if we hit this error memory server is in inconsistent state. 
We should start looking into Memory server error handling. One option is to not use the region ID again for this memory server instance and free the region id only during memory server exit

Reply:
Yes, we should go ahead and remove rest of the entries in case of an error. After de-registering the dataitems, should we return error to user in case one deregister fails ? Anyways region has already been destroyed.

Reply:
Since we already destroyed the region from nvmm, we need not return failure at this place.
anyway we need try and unregister all the dataitems and clear it from the region map.